File: cpp/math/matrix.cpp
// Multiply two fixed 3x3 integer matrices held as std::array of std::array.
// Build: c++ -std=c++20 -o matrix matrix.cpp

#include <array>
#include <cstddef>
#include <iostream>

namespace {

constexpr std::size_t size = 3;

using Matrix = std::array<std::array<int, size>, size>;

Matrix multiply(const Matrix& left, const Matrix& right)
{
    Matrix product{};

    for (std::size_t i = 0; i < size; ++i) {
        for (std::size_t j = 0; j < size; ++j) {
            for (std::size_t k = 0; k < size; ++k) {
                product[i][j] += left[i][k] * right[k][j];
            }
        }
    }
    return product;
}

int determinant(const Matrix& m)
{
    return m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1])
         - m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0])
         + m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
}

} // namespace

int main()
{
    constexpr Matrix left{{{2, -1, 0}, {1, 3, 4}, {0, 5, -2}}};
    constexpr Matrix right{{{1, 0, 2}, {-3, 1, 1}, {4, 2, 0}}};

    const Matrix product = multiply(left, right);

    for (const auto& row : product) {
        for (std::size_t j = 0; j < size; ++j) {
            std::cout << (j > 0 ? " " : "") << row[j];
        }
        std::cout << '\n';
    }
    std::cout << determinant(product) << '\n';
    return 0;
}

File: zig/math/matrix.zig
// Multiply two fixed 3x3 integer matrices held as arrays of arrays. (written against Zig 0.16)
// Build: zig build-exe matrix.zig

const std = @import("std");

const size = 3;
const Matrix = [size][size]i64;

const left: Matrix = .{ .{ 2, -1, 0 }, .{ 1, 3, 4 }, .{ 0, 5, -2 } };
const right: Matrix = .{ .{ 1, 0, 2 }, .{ -3, 1, 1 }, .{ 4, 2, 0 } };

fn multiply(a: Matrix, b: Matrix) Matrix {
    var product: Matrix = std.mem.zeroes(Matrix);
    for (0..size) |i| {
        for (0..size) |j| {
            for (0..size) |k| {
                product[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    return product;
}

fn determinant(m: Matrix) i64 {
    return m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]) -
        m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0]) +
        m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
}

pub fn main() void {
    const product = multiply(left, right);

    for (product) |row| {
        for (row, 0..) |value, index| {
            if (index > 0) std.debug.print(" ", .{});
            std.debug.print("{d}", .{value});
        }
        std.debug.print("\n", .{});
    }
    std.debug.print("{d}\n", .{determinant(product)});
}
